src/epoch.rs

```diff
@@ -7,77 +7,116 @@
-        let epoch = self.epoch.load(Relaxed);
-        ::std::sync::atomic::fence(SeqCst);
+    /// An epoch is internally represented as an integer that wraps around at some unspecified
+    /// point. The returned distance between two epochs is the minimum number of steps required to
+    /// go from `self` to `other` or go from `other` to `self`.
+    #[inline]
+    pub fn distance(&self, other: Self) -> usize {
+        let (e1, _) = self.decompose();
+        let (e2, _) = other.decompose();
+        cmp::min(e1.wrapping_sub(e2), e2.wrapping_sub(e1))
+    }
 
-        // Traverse the linked list of participant registries.
-        for participant in registries.iter(scope) {
-            match participant {
-                Err(IterError::LostRace) => {
-                    // We leave the job to the participant that won the race, which continues to
-                    // iterate the registries and tries to advance to epoch.
-                    return epoch;
-                }
-                Ok(local_epoch) => {
-                    let local_epoch = local_epoch.get();
-                    let (participant_is_pinned, participant_epoch) = local_epoch.get_state();
+    /// Returns `true` if the epoch is marked as pinned.
+    #[inline]
+    pub fn is_pinned(&self) -> bool {
+        self.decompose().1
+    }
 
-                    // If the participant was pinned in a different epoch, we cannot advance the
-                    // global epoch just yet.
-                    if participant_is_pinned && participant_epoch != epoch {
-                        return epoch;
-                    }
-                }
-            }
+    /// Returns the same epoch, but marked as pinned.
+    #[inline]
+    pub fn pinned(&self) -> Epoch {
+        Epoch {
+            data: self.data | 1,
         }
-        ::std::sync::atomic::fence(Acquire);
+    }
 
-        // All pinned participants were pinned in the current global epoch.  Try advancing the
-        // epoch. We increment by 2 and simply wrap around on overflow.
-        let epoch_new = epoch.wrapping_add(2);
-        self.epoch.store(epoch_new, Release);
-        epoch_new
+    /// Returns the same epoch, but marked as unpinned.
+    #[inline]
+    pub fn unpinned(&self) -> Epoch {
+        Epoch {
+            data: self.data & !1,
+        }
     }
-}
 
-impl Deref for Epoch {
-    type Target = AtomicUsize;
+    /// Returns the successor epoch.
+    ///
+    /// The returned epoch will be marked as pinned only if the previous one was as well.
+    #[inline]
+    pub fn successor(&self) -> Epoch {
+        Epoch {
+            data: self.data.wrapping_add(2),
+        }
+    }
 
-    fn deref(&self) -> &Self::Target {
-        &self.epoch
+    /// Decomposes the internal data into the epoch and the pin state.
+    #[inline]
+    fn decompose(&self) -> (usize, bool) {
+        (self.data >> 1, (self.data & 1) == 1)
```

Comment:
What about returning `(self.data, (self.data & 1) == 1)`?  Maybe it's a little bit more difficult interface to use, but may we can save an instruction :smiley:

Reply:
Well, it's just a shift operation that gets executed on cold paths, so it doesn't really matter... :)